### DISCUSSION TITLE
add boot parameters from UEFI

### DISCUSSION THREAD

**Discussion by tormodvolden** [memtest86plus/memtest86plus#418] [Q&A]
While debugging I run memtest in QEMU, and doc/HOW_TO_DEBUG_WITH_GDB.md explains how to run memtest automatically on boot, using `bcfg boot add 0 FS0:\EFI\boot\BOOT_X64.efi "memtest"`. This works fine. But how can I add a boot parameter like "usbdebug" to this? I have tried creating a file "options.txt" containing " usbdebug", and adding it with `bcfg boot -opt 0x0 FS0:\options.txt` but it doesn't take effect. I tried creating options.txt both as UTF-16 and ASCII. I also tried FS0:\memtext.efi instead of BOOT_X64.efi. I can see that the parameters have been added to the boot entry when I run `bcfg boot dump -v` but they are ignored.

**Comment by martinwhitaker** [comment]
I can't help much here, as I've never used the UEFI shell to load EFI images or used the `debug_memtest.sh` script. Memtest86+ does recognise boot options passed via the `LoadOptions` pointer associated with the EFI loaded image - I've tested that with both rEFInd and systemd-bootd. It expects to be passed a UTF-16 string (both rEFInd and systemd-bootd convert from ASCII automatically).

Personally, when debugging, I just change the default value of the options in the source code.

**Comment by tormodvolden** [comment]
Debugging config_init() I see that boot_params->cmd_line_ptr is set, but it points to a '\0' byte and boot_params->cmd_line_size is 1. Maybe it is a UEFI bug.

You mention it expects a UTF-16 string, but the code in config_init (with boot_params_addr being handed over in startup64.S) seems to deal with ASCII strings. Where has it been converted to ASCII? (Of course, this works all fine when entering the boot parameters as arguments when launching "memtest" at the UEFI shell prompt.)

**Reply by martinwhitaker** [reply]
The EFI load options are converted to ASCII in `get_cmd_line()` in  `efisetup.c`.

**Reply by tormodvolden** [reply]
Thanks! When I saw "truncate the command line if we find a non-ASCII character" in efisetup.c I remembered having read the note "Add extra spaces in the beginning of the line in the file. There is a [byte order mark](https://en.wikipedia.org/wiki/Byte_order_mark) at the beginning of the line that will squash any character next to it which will cause an error when booting." from https://wiki.archlinux.org/title/EFISTUB#bcfg when I was fighting with the bcfg boot -opt command. PR #421.
